anvil/sample.py

```diff
@@ -293,14 +301,17 @@ def sample(
     rejected = n_batches * batch_size - accepted
     fraction = accepted / (accepted + rejected)
 
-    log.debug(f"Accepted: {accepted}, Rejected: {rejected}, Fraction: {fraction:.2g}")
+    log.info(f"Accepted: {accepted}, Rejected: {rejected}, Fraction: {fraction:.2g}")
     log.debug(f"Returning a decorrelated chain of length: {actual_length}")
-    return decorrelated_chain
 
+    #np.savetxt("ensemble_out.txt", decorrelated_chain)
+    with open("acceptance.txt", "a") as f:
+        f.write(f"{fraction}\n")
 
-_sample_training_output = collect("sample", ("training_context",))
+    return decorrelated_chain
 
+_sample_training_output = collect("sample", ("training_context",))
```

Comment:
I don't like this 'collect over input card' because the values in the input card are used rather than being parsed by `ConfigParser` (production functions work, however). 

Apart from skipping all the checks, this means we cannot return anything other than the input in a parse function. For example, if I let `def produce_target_dist(self, target: str)` become `def parse_target(self, target: str)` and then use `target` in place of `target_dist` throughout the code (which is much nicer) then the training works but the sampling does not, since `_sample_training_output` sees the string version of `target` provided in the training input card, not the object returned by `parse_target`.

Reply:
So moving forwards I am looking at different ways to ensure consistency between the training, sampling and report-making stages without using 'collect over input card'.

Reply:
Not that you would know this but reportengine has a `parse_from_` mechanism which means in the runcard I can do:

```
sample: {from_: training_context}
```

but it also means inside a production rule you can do something like
```
self.parse_from_(namespace, key)
```

I'd have to see the exact requirement to give a  better example but I think this would solve you're problem because `parse_from_` would return the output of the parsing function acting on the key so I think you could get your consistency.

Reply:
The first example there is assuming sample isn't a production rule. It's all a bit complicated but I think it would be best for me to show you a proper example of this in action

Reply:
Ah excellent, yes I think I understand vaguely how this could work, based on the 'Validphys 2 guide'

Reply:
The documentation really is quite poor for some of these features. Soon the nnpdf code will be public so you will be able to see examples of using reportengine including using `parse_from_`. Since `training_output` has an `as_input()` method you can actually write:

`self.parse_from_:(training_output, <key>)` which is what I really should have done in the `training_context` but for the selection of keys we wanted. I was incredibly lazy so just dump the unprocessed config instead which is really rubbish. Sorry

Reply:
I guess it was because we were taking so many of the keys. But a bit of verbosity is a small price to pay for this thing to work properly.

Reply:
no need to apologise! It's often been the case that I've had funny ideas about how to do things and you've had to quickly morph them into something that works with reportengine.. Anyway this was always gonna be a case of iterative improvement and it's getting there..